### PR TITLE
Week 3 Game Logic

### PR DESCRIPTION
**Pull request by jacobmcphail**
Features Added:
- Hitboxes
- Game/Object Management
- Collision Handling
- Basic Inventory System
- Player Aiming
- Partial Turret placement/pickup (In progress)
- A whole lot of object classes
- Audio Manager

Note: There was some code redundancy/duplication created during this week. Will be gone before the end of week 4.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # 4981_Linux_Game
 
-Brace yourself for the thrill of zombie-fighting in space! *Name TBD* Features a visually stunning 2D top-down network arena style tower defense game parallel with unmatched networking speeds that will blow your mind. All this is possible via the finesse of CST's Finest.  
+Brace yourself for the thrill of zombie-fighting in space! Space Creep Features a visually stunning 2D isometric network arena style defense game parallel with unmatched networking speeds that will blow your mind. All this is possible via the finesse of CST's Finest.
```
Comment (isaacmorneau) [inline README.md]: Ill merge this regardless but there was a name change to MAR-Z

File: src/Base.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
+#include "Base.h"
+#include "Marine.h"
+#include<random>
+
+Base::Base() {
+	this->setX((MAP_WIDTH/2)-BASE_WIDTH);
+	this->setY((MAP_HEIGHT/2)-BASE_HEIGHT);
```
Comment (isaacmorneau) [inline src/Base.cpp]: For parameters please try and keep parts spaced out as per the style guide
`this->setX((MAP_WIDTH/2)-BASE_WIDTH);`
should be
`this->setX((MAP_WIDTH / 2) - BASE_WIDTH);`

File: src/Base.cpp [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	int gab = 100;
+	
+
+	switch(distr(eng)){
+		case 0://North
+	
+			//y point is fixed
+			y = this->getY()-gab;
+			//x point is randomly generated;	
+			x = distr(eng, decltype(distr)::param_type(this->getX()-gab, this->getX()+this->getWidth()+gab));
```
Comment (isaacmorneau) [inline src/Base.cpp]: Same as above, don't be afraid to add white space or even break it into multiple lines for readability

File: src/ShotGun.cpp [hunk 4]
```diff
@@ -0,0 +1,21 @@
+/*
+    Created by Maitiu Morton 2/8/2017
+*/
+#include "ShotGun.h"
+
+ShotGun::ShotGun(){
```
Comment (isaacmorneau) [inline src/ShotGun.cpp]: Why does this not use member initializers?

File: src/Weapon.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    void shot();//decreases the amount of ammo
+    int getClipMax(); //returns max amount clip can hold
+    std::string getType();
+    int getReloadSpeed();
+
+	Weapon();
+	~Weapon();
+
+protected:
+    std::string type = "no type";
```
Comment (isaacmorneau) [inline src/Weapon.h]: defaults should be set as parameters in the constructor not as static initialization in headers

File: src/Zombie.cpp [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+void Zombie::generateRandomMove(){
+	int tx = this->getX()+50;
+	int ty = this->getY()+50;
+	if(tx<=0 || tx>=2000 || ty<=0 || ty>=2000){
+		this->setAngle(this->getRandomAngle());
+	}
+		
+	double cosVal = cos(this->getAngle()*PI/180.0);
```
Comment (isaacmorneau) [inline src/Zombie.cpp:121]: why are we using _degrees_ ? _radians_ are what all the math functions use and return.
It's inefficient to calculate the conversion every single time.

if you want a refresher i'd be happy to give you one.

File: src/Turret.cpp [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+bool Turret::ammoCheckTurret() {
+    if (ammo > 0)
+        return true;
+    else
+        return false;
+}
+	
+ // returns true if turret has >=1 health, false otherwise  
+bool Turret::healthCheckTurret() {
+    if (health > 0)
```
Comment (isaacmorneau) [inline src/Turret.cpp]: Seriously? why don't we just go 
`return (health > 0);`
it's clear and doesn't make anyone cringe when reading it.
I'll even add it to the style guide so everyone knows it's ok.

File: src/CollisionHandler.cpp [hunk 8]
```diff
@@ -0,0 +1,57 @@
+#include "CollisionHandler.h"
+
+CollisionHandler::CollisionHandler() {
+	this->quadtreeMov = new Quadtree(0, {0,0,2000,2000});
+	this->quadtreePro = new Quadtree(0, {0,0,2000,2000});
+	this->quadtreeDam = new Quadtree(0, {0,0,2000,2000});
+}
+
+CollisionHandler::~CollisionHandler() {
```
Comment (isaacmorneau) [inline src/CollisionHandler.cpp:9]: props to @JohnAgapeyev  for catching this

there's a memory leak as the destructor never frees the quadtreeMov  quadtreePro  quadtreeDam

Comment (jacobmcphail) [reply src/CollisionHandler.cpp:9]: Ah good catch thanks

File: src/GameStateMatch.cpp [hunk 9]
```diff
@@ -25,16 +25,63 @@ bool GameStateMatch::load() {
...
+	success = GameManager::instance()->createMarine(this->game->renderer, 1500, 1500);
+
+
+	Turret* dumbTurret = GameManager::instance()->getTurret(GameManager::instance()->createTurret());
+	if (!dumbTurret->texture.loadFromFile("assets/texture/turret.png", this->game->renderer)) {
+		printf("Failed to load the player texture!\n");
+		success = false;
+	}
+
+	Zombie* dumbZombie = new Zombie();
```
Comment (isaacmorneau) [inline src/GameStateMatch.cpp]: 45 to 51 two assignments yet the old zombie is never freed
this is a memory leak

File: src/GameStateMatch.cpp [hunk 10]
```diff
@@ -25,16 +25,63 @@ bool GameStateMatch::load() {
...
+	GameManager::instance()->addZombie(dumbZombie);
+	if (!dumbZombie->texture.loadFromFile("assets/texture/zombie.png", this->game->renderer)) {
+		printf("Failed to load the player texture!\n");
+		success = false;
+	}
+	dumbZombie = new Zombie();
```
Comment (MikeGoll) [inline src/GameStateMatch.cpp]: Lines 51 - 56 are redundant. You are assigning a new pointer to an already existing object and assigning it the exact same texture.

This is also a memory leak as the new object pointer is never freed.

File: src/Weapon.cpp [hunk 11]
```diff
@@ -1,7 +1,21 @@
 #include "Weapon.h"
 
-Weapon::Weapon(){
-
+Weapon::Weapon(std::string type,
```
Comment (isaacmorneau) [inline src/Weapon.cpp]: So close! you fixed all sub classes but this should use member initializers as well.

**Pull request merged**